File: aufgabe_1_security/csharp/DatabaseProxy.cs
using System;

namespace ProxyPatternExercise1
{
    public interface IDatabase {
        void DeleteUser(string username);
    }

    public class RealDatabase : IDatabase {
        public void DeleteUser(string username) {
            Console.WriteLine($"User {username} wurde aus der Datenbank gelöscht!");
        }
    }

    public class DatabaseProxy : IDatabase {
        private RealDatabase _realDatabase;
        private string _userRole;

        public DatabaseProxy(string userRole) {
            // TODO: Speichere userRole in _userRole
            // TODO: Initialisiere _realDatabase = new RealDatabase();
        }

        public void DeleteUser(string username) {
            // TODO: Prüfe, ob _userRole == "admin" ist
            // Wenn JA: Rufe _realDatabase.DeleteUser auf
            // Wenn NEIN: Console.WriteLine("Fehler: Zugriff verweigert!");
        }
    }

    /* TEST BEREICH (Kopiere dies in eine Main Methode)
    IDatabase gastProxy = new DatabaseProxy("gast");
    gastProxy.DeleteUser("max_mustermann"); // Sollte "Zugriff verweigert" ausgeben

    IDatabase adminProxy = new DatabaseProxy("admin");
    adminProxy.DeleteUser("max_mustermann"); // Sollte den User löschen
    */
}

File: aufgabe_1_security/python/database_proxy.py
from abc import ABC, abstractmethod

class DatabaseInterface(ABC):
    @abstractmethod
    def delete_user(self, username: str):
        pass

class RealDatabase(DatabaseInterface):
    def delete_user(self, username: str):
        print(f"User {username} wurde aus der Datenbank gelöscht!")

class DatabaseProxy(DatabaseInterface):
    def __init__(self, user_role: str):
        # TODO: Speichere die Rolle (user_role) in self.user_role
        # TODO: Erstelle eine Instanz von RealDatabase in self.real_database
        pass

    def delete_user(self, username: str):
        # TODO: Prüfe, ob self.user_role == "admin" ist
        # Wenn JA: Rufe delete_user auf self.real_database auf
        # Wenn NEIN: printe "Fehler: Zugriff verweigert!"
        pass

# --- TEST BEREICH ---
# gast_proxy = DatabaseProxy("gast")
# gast_proxy.delete_user("max_mustermann") # Sollte "Zugriff verweigert" ausgeben

# admin_proxy = DatabaseProxy("admin")
# admin_proxy.delete_user("max_mustermann") # Sollte den User löschen
